### PR TITLE
Add support for RFC Draft12

### PR DESCRIPTION
**Pull request by Maximus-**
Support the changes made in RFC Draft12, specifically in the four-pass encryption scheme. 
I've ran all the tests and they still succeed. 

There's possibly some cleaning up that needs doing, but largely it's functional. Let me know if there's any stylistic, functional, or design changes you'd like made. Also need to sign the CLA. :)

### CODE REVIEW

File: quic_lb.c [hunk 1]
```diff
@@ -89,59 +93,140 @@ quic_lb_pcid_decrypt(void *ctx, void *cid, void *sid, size_t *cid_len)
...
-    return ERR_REJECT;
 }
 
 static void
 quic_lb_scid_encrypt(void *ctx, void *cid)
 {
     struct quic_lb_server_ctx *cfg = ctx;
-    UINT8  *sid = (UINT8 *)cid + 1, *nonce = sid + cfg->sidl;
+
+    size_t sidl = cfg->sidl + cfg->nonce_len;
```
Comment (martinduke) [inline quic_lb.c:186]: This is a weird name. Maybe total_len?

File: quic_lb.c [hunk 2]
```diff
@@ -400,3 +499,49 @@ quic_lb_decrypt_cid(void *ctx, void *cid, void *sid, size_t *cid_len)
             context->decrypt(ctx, cid, sid, cid_len));
 }
 
+void
```
Comment (martinduke) [inline quic_lb.c:502]: Is this being exercised somehow?

**Review by martinduke — APPROVED** [review]
By inspection, this seems to be correct, minus a few nits.

There are a couple of more structural issues that would be good to pursue in further PRs:
1) The current test suite just tests for self-consistency (we can decode our own encodings). It would be good to have new tests that correspond to the example in 4.3.2 and the test vectors in Appendix B.

2) The algorithms have converged enough that I don't think we need algorithm identifiers or separate functions for each encrypt/decrypt. For example, the encrypt function could be:

if (!cfg) return GenerateRandomCID();
cid = WriteSidAndNonce();
if (!cfg->crypto_ctx) return cid;
if (cfg->nonce_len + cfg->sidl == 16) {
  BlockEncrypt(cid);
  return cid;
}
FourPassEncryption(cid);
return cid;

**Comment by martinduke** [comment]
Fixes #13

File: quic_lb.c (hunk 2 above)
Comment (Maximus-) [reply quic_lb.c:502]: It was - it seems it got lost in transit. Addressing :)

**Comment by Maximus-** [comment]
>The current test suite just tests for self-consistency (we can decode our own encodings). It would be good to have new tests that correspond to the example in 4.3.2 and the test vectors in Appendix B.

I was testing during development against RFC values - but If I recall correctly, it was hard to set them up in the tests. Will look into it before completing this PR :)

File: quic_lb.c (hunk 1 above)
Comment (Maximus-) [reply quic_lb.c:186]: Ack

**Comment by Maximus-** [comment]
Addressed most of the concerns above

Remaining items for subsequent PR:
- Structural changes, remove SCID/BCID
- Add tests to use RFC values

**Pull request merged**